@Nested tests in a superclass do not work with test profiles

### Describe the bug

If `@Nested` tests are declared in a shared superclass, and if 2 or more test subclasses exist, any test class using test profiles will crash with an `AmbiguousResolutionException`.

The relevant portions of the stack trace are:

```
org.junit.jupiter.api.extension.TestInstantiationException: Failed to create test instance
	at io.quarkus.test.junit.QuarkusTestExtension.initTestState(QuarkusTestExtension.java:786)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:752)
    ...
Caused by: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at io.quarkus.runner.bootstrap.RunningQuarkusApplicationImpl.instance(RunningQuarkusApplicationImpl.java:90)
	at io.quarkus.test.junit.QuarkusTestExtension.createActualTestInstance(QuarkusTestExtension.java:792)
	at io.quarkus.test.junit.QuarkusTestExtension.initTestState(QuarkusTestExtension.java:773)
	... 76 more
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.runner.bootstrap.RunningQuarkusApplicationImpl.instance(RunningQuarkusApplicationImpl.java:88)
	... 78 more
Caused by: jakarta.enterprise.inject.AmbiguousResolutionException: Beans: [CLASS bean [class=org.acme.Nested2Test, id=3b65ca95454361c99a346a4310722038d56f59c8], CLASS bean [class=org.acme.Nested1Test, id=570d58f8116768968fb960c30fc6d2e6df1364a9]]
	at io.quarkus.arc.impl.InstanceImpl.bean(InstanceImpl.java:291)
	at io.quarkus.arc.impl.InstanceImpl.getInternal(InstanceImpl.java:309)
	at io.quarkus.arc.impl.InstanceImpl.get(InstanceImpl.java:190)
	... 83 more
```

There is a workaround though, see reproducer below, so I wouldn't consider this bug as an urgent one.

### Expected behavior

All tests should run and pass.

### Actual behavior

Any test class annotated with `@TestProfile` will throw `AmbiguousResolutionException` when executed.

### How to Reproduce?

The following is a small reproducer. Running `Nested1Test` will exhibit the faulty behavior:

```java
public class AbstractNestedTest {
    @Nested
    class NestedTests {
        @Test
        public void dummyNestedTest() {
        }
    }
}

```

```java
@QuarkusTest
@TestProfile(Nested1Test.Profile.class)
public class Nested1Test extends AbstractNestedTest {

    public static class Profile implements QuarkusTestProfile {
    }

    @Test
    public void dummyTest1() {
    }

    // Uncomment the following for a workaround:
    // @Nested
    // class NestedTests extends AbstractNestedTest.NestedTests {
    // }
}
```

```java
@QuarkusTest
public class Nested2Test extends AbstractNestedTest {
    @Test
    public void dummyTest2() {
    }
}
```

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

openjdk version "17.0.6" 2023-01-17 OpenJDK Runtime Environment Temurin-17.0.6+10 (build 17.0.6+10) OpenJDK 64-Bit Server VM Temurin-17.0.6+10 (build 17.0.6+10, mixed mode)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.2.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

I think it's a class loading issue. The if expression below:

https://github.com/quarkusio/quarkus/blob/752ec72787e8c67f213bd8b4ee65f616e113da2f/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java#L769 

...fails, but it shouldn't, because the 2 classes were loaded by different class loaders.

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/CommonNestedTest.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/CommonNestedTest.java
@@ -1,0 +1,27 @@
+package io.quarkus.it.main;
+
+import static org.hamcrest.Matchers.is;
+
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+import io.restassured.RestAssured;
+
+public abstract class CommonNestedTest {
+
+    public String defaultProfile() {
+        return "Hello";
+    }
+
+    @Nested
+    class NestedTests {
+        @Test
+        public void testProfileFromNested() {
+            RestAssured.when()
+                    .get("/greeting/Stu")
+                    .then()
+                    .statusCode(200)
+                    .body(is(defaultProfile() + " Stu"));
+        }
+    }
+}

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/QuarkusTestNested1WithCommonCaseTest.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/QuarkusTestNested1WithCommonCaseTest.java
@@ -1,0 +1,14 @@
+package io.quarkus.it.main;
+
+import io.quarkus.test.junit.QuarkusTest;
+import io.quarkus.test.junit.TestProfile;
+
+@QuarkusTest
+@TestProfile(QuarkusTestNestedWithTestProfileTestCase.OuterProfile.class)
+public class QuarkusTestNested1WithCommonCaseTest extends CommonNestedTest {
+
+    @Override
+    public String defaultProfile() {
+        return "OuterProfile";
+    }
+}

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/QuarkusTestNested2WithCommonCaseTest.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/QuarkusTestNested2WithCommonCaseTest.java
@@ -1,0 +1,7 @@
+package io.quarkus.it.main;
+
+import io.quarkus.test.junit.QuarkusTest;
+
+@QuarkusTest
+public class QuarkusTestNested2WithCommonCaseTest extends CommonNestedTest {
+}

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/AbstractJvmQuarkusTestExtension.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/AbstractJvmQuarkusTestExtension.java
@@ -13,12 +13,14 @@ import java.util.Collection;
 import java.util.Deque;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Optional;
 import java.util.function.Consumer;
 import java.util.stream.Collectors;
 
 import jakarta.enterprise.inject.Alternative;
 
 import org.jboss.jandex.Index;
+import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.extension.ExtensionContext;
 
 import io.quarkus.bootstrap.BootstrapConstants;
@@ -213,7 +215,42 @@ public class AbstractJvmQuarkusTestExtension extends AbstractQuarkusTestWithCont
     }
 
     protected Class<? extends QuarkusTestProfile> getQuarkusTestProfile(ExtensionContext extensionContext) {
-        Class<?> testClass = extensionContext.getRequiredTestClass();
+        // If the current class or any enclosing class in its hierarchy is annotated with `@TestProfile`.
+        Class<? extends QuarkusTestProfile> testProfile = findTestProfileAnnotation(extensionContext.getRequiredTestClass());
+        if (testProfile != null) {
+            return testProfile;
+        }
+
+        // Otherwise, if the current class is annotated with `@Nested`:
+        if (extensionContext.getRequiredTestClass().isAnnotationPresent(Nested.class)) {
+            // let's try to find the `@TestProfile` from the enclosing classes:
+            testProfile = findTestProfileAnnotation(extensionContext.getRequiredTestClass().getEnclosingClass());
+            if (testProfile != null) {
+                return testProfile;
+            }
+
+            // if not found, let's try the parents
+            Optional<ExtensionContext> parentContext = extensionContext.getParent();
+            while (parentContext.isPresent()) {
+                ExtensionContext currentExtensionContext = parentContext.get();
+                if (currentExtensionContext.getTestClass().isEmpty()) {
+                    break;
+                }
+
+                testProfile = findTestProfileAnnotation(currentExtensionContext.getTestClass().get());
+                if (testProfile != null) {
+                    return testProfile;
+                }
+
+                parentContext = currentExtensionContext.getParent();
+            }
+        }
+
+        return null;
+    }
+
+    private Class<? extends QuarkusTestProfile> findTestProfileAnnotation(Class<?> clazz) {
+        Class<?> testClass = clazz;
         while (testClass != null) {
             TestProfile annotation = testClass.getAnnotation(TestProfile.class);
             if (annotation != null) {

